source/common/unified_memory_allocation/src/memory_tracker.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ */
+
+#include "memory_tracker_internal.h"
+#include <uma/memory_provider.h>
+#include <uma/memory_provider_ops.h>
+
+#include <map>
+#include <mutex>
+#include <shared_mutex>
+
+// TODO: reimplement in C and optimize...
+struct uma_memory_tracker_t {
+    enum uma_result_t add(void *pool, const void *ptr, size_t size) {
+        std::unique_lock<std::shared_mutex> lock(mtx);
+
+        auto ret =
+            map.try_emplace(reinterpret_cast<uintptr_t>(ptr), size, pool);
+        return ret.second ? UMA_RESULT_SUCCESS : UMA_RESULT_ERROR_UNKNOWN;
+    }
+
+    enum uma_result_t remove(const void *ptr, size_t size) {
+        std::unique_lock<std::shared_mutex> lock(mtx);
+
+        map.erase(reinterpret_cast<uintptr_t>(ptr));
+
+        // TODO: handle removing part of the range
+        (void)size;
+
+        return UMA_RESULT_SUCCESS;
+    }
+
+    void *find(const void *ptr) {
+        std::shared_lock<std::shared_mutex> lock(mtx);
+
+        auto intptr = reinterpret_cast<uintptr_t>(ptr);
+        auto it = map.upper_bound(intptr);
+        if (it == map.begin()) {
+            return nullptr;
+        }
+
+        --it;
+        if (intptr >= it->first && intptr < it->first + it->second.first) {
```

Comment:
this looks like magic if you don't know what you are looking at. Maybe create more useful variable names or add comments?
`size = it->first +it->second.first`

Reply:
Done.